### DISCUSSION TITLE
Putchar Default

### DISCUSSION THREAD

**Discussion by thestumbler** [modm-io/modm#1089] [General]
I noticed in commit 93af956745849d5bb2626c55f68d0fba20b69405 that `putchar_()` function was included by default. That definition coincides with my current usage, but I do have plans to separate PRINTF serial port from the logging port down the road ( I suspect I would need to add another PRINTF function to accomplish this, such as [here](https://embeddedartistry.com/blog/2019/11/06/an-embedded-friendly-printf-implementation/) ).

Can / should the built-in PUTCHAR_ function be made weak somehow, so it could be replaced if an application needs it?

EDIT: I also just realized that while PUTCHAR_ was added to the CPP file, it was not added to the HPP file. Was that intentional?

**Comment by salkinium** [comment]
We patched the original [printf implementation to make `putchar_` weak by default](https://github.com/modm-ext/printf-partial/blob/72716295bcaa14f96ad9725b40dcc9dff0930eac/src/printf/printf.c#L298-L301). This prevented me from declaring the BSP `putchar_` as weak as well.
Perhaps this was the wrong choice, and we should just not default it and then there will just be linker errors when you try to use printf, which would probably be better than just silently discarding the printf output… 😬

The workaround is to use a custom BSP of course, where you can define it however you want. Note that you can create your own IOStreams with a custom IODeviceWrapper and use the `stream.printf()` method, which then calls the same printf implementation but with a different output. The libc `printf` function is basically only there for integration of third-party code.

**Reply by thestumbler** [reply]
That makes sense. I'm not sure how to make a BSP which changes PUTCHAR, however, since the definition is in a top-level BOARD.CPP.IN and not in the individual board directories. Does one just make a custom BOARD.CPP in their BSP directory? If so, I don't see any of the example projects doing that.

**Reply by salkinium** [reply]
> Does one just make a custom BOARD.CPP in their BSP directory?

Yes, and then a regular `env.copy` of the local file instead of the top-level template.
